### PR TITLE
feat(notification-service): Implementation of notification grouping feature

### PR DESCRIPTION
**Pull request by karuneshsf**
## Description
Implemented notification grouping feature as follows

1. added feature to group many notification to one notification
2. added feature to have sleep time for recipient(s)
3. added feature to draft notification
4. added feature to send notification without drafting if it is critical
5. added feature to group many notification as one notification

Fixes # (issue)
GH-1806
## Type of change
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?
- [x] Performed a self-review of my own code
- [x] npm test passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/cli/README.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ $ npm install -g @sourceloop/cli
 $ sl COMMAND
 running command...
 $ sl (-v|--version|version)
-@sourceloop/cli/6.0.0 linux-x64 node-v18.19.0
+@sourceloop/cli/6.0.0 darwin-arm64 node-v16.15.0
```
Comment (samarpan-b) [inline packages/cli/README.md:21]: why this change ?

File: .vscode/settings.json [hunk 2]
```diff
@@ -3,5 +3,8 @@
   "editor.formatOnSave": true,
   "editor.codeActionsOnSave": {
     "source.organizeImports": true
-  }
-}
+  },
+  "cSpell.words": [
+    "Notif"
```
Comment (samarpan-b) [inline .vscode/settings.json:8]: why ?

File: packages/core/locales/en.json [hunk 3]
```diff
@@ -169,5 +169,24 @@
...
 	"No factory found for sf.passport.strategyFactory.localPassword": "No factory found for sf.passport.strategyFactory.localPassword",
-	"No factory found for sf.passport.strategyFactory.otpAuth": "No factory found for sf.passport.strategyFactory.otpAuth"
+	"No factory found for sf.passport.strategyFactory.otpAuth": "No factory found for sf.passport.strategyFactory.otpAuth",
+	"The key 'repositories.UserNotificationSettingsRepository' is not bound to any value in context RequestContext-NJzgVogAQu2g51aqkFd2OA-92 (context": " RequestContext-NJzgVogAQu2g51aqkFd2OA-92, binding: repositories.UserNotificationSettingsRepository, resolutionPath: controllers.NotificationController --> @NotificationController.constructor[5] --> services.ProcessNotificationService --> @ProcessNotificationService.constructor[4] --> sourceloop.services.notification.notificationSettingFilter --> @NotificationUserSettingsProvider.constructor[0])",
```
Comment (samarpan-b) [inline packages/core/locales/en.json:173]: revert the changes in this file please

File: sandbox/chat-notification-pubnub-example/services/notifications-service/src/application.ts [hunk 4]
```diff
@@ -2,29 +2,29 @@
 //
 // This software is released under the MIT License.
 // https://opensource.org/licenses/MIT
-import {BootMixin} from '@loopback/boot';
-import {ApplicationConfig} from '@loopback/core';
-import {RepositoryMixin} from '@loopback/repository';
-import {RestApplication} from '@loopback/rest';
+import { BootMixin } from '@loopback/boot';
```
Comment (samarpan-b) [inline sandbox/chat-notification-pubnub-example/services/notifications-service/src/application.ts:5]: Fix please

File: services/notification-service/migrations/sqls/20231128083406-alter-db-to-support-multiple-notification-as-one-notification-up.sql [hunk 5]
```diff
@@ -0,0 +1,35 @@
+ALTER TABLE main.notifications
+ADD COLUMN IF NOT EXISTS  is_draft boolean default FALSE;
```
Comment (samarpan-b) [inline services/notification-service/migrations/sqls/20231128083406-alter-db-to-support-multiple-notification-as-one-notification-up.sql:2]: why do we need this is_draft column ?

File: services/notification-service/migrations/sqls/package.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (samarpan-b) [inline services/notification-service/migrations/sqls/package.json:1]: revert pls

File: services/notification-service/src/controllers/user-notification-settings.controller.ts [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+            exclude: ['id'],
+          }),
+        },
+      },
+    })
+    userNotificationSettings: Omit<UserNotificationSettings, 'id'>,
+    @inject(AuthenticationBindings.CURRENT_USER)
+    currentUser: IAuthUserWithPermissions,
+  ): Promise<UserNotificationSettings> {
+    if (currentUser.id !== userNotificationSettings.userId) {
```
Comment (samarpan-b) [inline services/notification-service/src/controllers/user-notification-settings.controller.ts:68]: remove this check for now

File: services/notification-service/src/controllers/user-notification-settings.controller.ts [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+  async deleteById(
+    @param.path.string('id') id: string,
+    @inject(AuthenticationBindings.CURRENT_USER)
+    currentUser: IAuthUserWithPermissions,
+  ): Promise<void> {
+    await this._verifyOwned(id, currentUser);
+    await this.userNotificationSettingsRepository.deleteById(id);
+  }
+
+  private async _verifyOwned(
```
Comment (samarpan-b) [inline services/notification-service/src/controllers/user-notification-settings.controller.ts:245]: No need to check for current user for now

**Comment by yeshamavani** [comment]
PR title

File: services/notification-service/src/providers/notification-user-settings.service.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            const element = toUser;
+            const currentTime = new Date();
+            const sleepTime = await this.userNotifSettingsRepository.findOne({
+              where: {
+                userId: element.id,
+                sleepStartTime: {lte: currentTime},
+                sleepEndTime: {gte: currentTime},
+              },
+            });
+            element.isDraft = false;
```
Comment (samarpan-b) [inline services/notification-service/src/providers/notification-user-settings.service.ts:56]: This should be in notif-user table

File: services/notification-service/src/services/process-notification.service.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+import {Notification} from '../models';
+import {
+  NotificationRepository,
+  NotificationUserRepository,
+} from '../repositories';
+import {
+  INotificationSettingFilterFunc,
+  INotificationUserManager,
+} from '../types';
+const maxBodyLen = 1000;
```
Comment (samarpan-b) [inline services/notification-service/src/services/process-notification.service.ts:17]: move this to process.env

File: services/notification-service/src/services/process-notification.service.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+      );
+    }
+    //Cutting the length of the notification to predefined max length only in case if it is not a grouped notification
+    if (
+      (!notification.isGrouped || !notification.groupKey) &&
+      notification.body.length > maxBodyLen
+    ) {
+      notification.body = notification.body.substring(0, maxBodyLen - 1);
+    }
+    notif = await this.insertDataInDb(notification);
```
Comment (samarpan-b) [inline services/notification-service/src/services/process-notification.service.ts:62]: Use transactions. Do rollback if anything fails

File: services/notification-service/src/services/process-notification.service.ts [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    if (!notification.isDraft) {
+      notif = await this.insertDataInDb(notification);
```
Comment (samarpan-b) [inline services/notification-service/src/services/process-notification.service.ts:64]: This seems to be happening twice

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 13]
```diff
@@ -2,6 +2,6 @@
   "markdown.extension.toc.levels": "2..2",
   "editor.formatOnSave": true,
   "editor.codeActionsOnSave": {
-    "source.organizeImports": true
+    "source.organizeImports": "explicit"
```
Comment (samarpan-b) [inline .vscode/settings.json:5]: revert this

File: sandbox/chat-notification-pubnub-example/services/notifications-service/package.json [hunk 14]
```diff
@@ -59,8 +59,8 @@
     "@loopback/rest": "^13.1.0",
     "@loopback/rest-explorer": "^6.1.3",
     "@loopback/service-proxy": "^6.1.3",
-    "@sourceloop/core": "^10.1.0",
-    "@sourceloop/notification-service": "^9.0.1",
+    "@sourceloop/core": "^10.0.0",
```
Comment (samarpan-b) [inline sandbox/chat-notification-pubnub-example/services/notifications-service/package.json:62]: revert pls

File: services/notification-service/package.json [hunk 15]
```diff
@@ -29,7 +29,10 @@
     "clean": "lb-clean dist *.tsbuildinfo .eslintcache",
     "coverage": "nyc npm run test",
     "prepublishOnly": "npm run test",
-    "postinstall": "node migration.js"
+    "postinstall": "node migration.js",
+    "db:migrate": "../../node_modules/db-migrate/bin/db-migrate up --config './database.json'",
```
Comment (samarpan-b) [inline services/notification-service/package.json:33]: revert this please

File: services/notification-service/src/__tests__/acceptance/helper.ts [hunk 16]
```diff
@@ -49,3 +49,7 @@ export interface AppWithClient {
   app: NotificationApplication;
   client: Client;
 }
+
+export function setEnvVars() {
+  process.env.TEST_ENV = '1';
```
Comment (samarpan-b) [inline services/notification-service/src/__tests__/acceptance/helper.ts:54]: process.env.NODE_ENV = 'test';

File: services/notification-service/src/controllers/notification.controller.ts [hunk 17]
```diff
@@ -94,12 +111,216 @@ export class NotificationController {
...
   }
 
+  @authenticate(STRATEGY.BEARER)
+  @authorize({
+    permissions: [
+      PermissionKey.CreateNotification,
+      PermissionKey.CreateNotificationNum,
+    ],
+  })
+  @post(`${basePath}/group/{groupKey}`, {
```
Comment (samarpan-b) [inline services/notification-service/src/controllers/notification.controller.ts:126]:
```suggestion
  @post(`${basePath}/groups/{groupKey}`, {
```

File: services/notification-service/src/services/process-notification.service.ts [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+    for (const element of notifications) {
+      groupNotificationBody += element.body;
+      if (element.id) {
+        //groupedNotifIds are the notification ids to which notification has been sent, and the isDraft status to be updated to be false
+        groupedNotifIds.push(String(element.id));
+      }
+    }
+    notification.body = groupNotificationBody;
+    const notifTx =
+      process.env.TEST_ENV !== '1'
```
Comment (samarpan-b) [inline services/notification-service/src/services/process-notification.service.ts:63]:
```suggestion
      process.env.NODE_ENV !== 'test'
```

**Comment by yeshamavani** [comment]
apart from comments looks good to me

**Comment by samarpan-b** [comment]
CI is failing

**Review by samarpan-b — CHANGES_REQUESTED** [review]

**Review by samarpan-b — APPROVED** [review]

**Review by yeshamavani — APPROVED** [review]

**Pull request merged**